lib/core/src/Cardano/Wallet/Primitive/Types.hs

```diff
@@ -491,17 +499,45 @@ instance NFData Address
 instance Buildable Address where
     build = build . toText
 
+instance ToText Address where
+    toText = T.decodeUtf8
+        . convertToBase Base16
+        . getAddress
+
 instance FromText Address where
-    fromText x = maybe
-        (Left $ TextDecodingError err)
-        (pure . Address)
-        (decodeBase58 bitcoinAlphabet $ T.encodeUtf8 x)
+    fromText = bimap textDecodingError Address
+        . convertFromBase Base16
+        . T.encodeUtf8
       where
-        err = "Unable to decode Address: expected Base58 encoding"
+        textDecodingError = TextDecodingError . show
 
-instance ToText Address where
-    toText = T.decodeUtf8 . encodeBase58 bitcoinAlphabet . getAddress
+-- | An abstract class to allow encoding of addresses depending on the target
+-- backend used.
+class EncodeAddress t where
+    encodeAddress :: Proxy t -> Address -> Text
 
+-- | An abstract class to allow decoding of addresses depending on the target
+-- backend used.
+class DecodeAddress t where
+    decodeAddress :: Proxy t -> Text -> Either TextDecodingError Address
```

Comment:
I wonder if we constrain ourselves here by requiring expressive bech32 errors to live in `TextDecodingError`… or maybe it won't matter in the end

Reply:
I don't see what would be the alternative?

Here we are stating that decoding could fail and we are tagging the error with `TextDecodingError`. For other future backend implementation, regardles beach32 or other encoding, we will do the same - wrap the error with some error constructor.

I probably don't see the concern you are noticing - can you expand on this?

Reply:
@Anviking I won't matter. The `TextDecodingError` is really just a wrapper around a user-friendly error message from parsing a `Text` value, which is, exactly what we do / want here. This also makes it fairly easy to re-use existing handlers `eitherToParser`and other utilities built on top of the `FromText` class.